commerce/domain/pricing/cart/PricingBasicCalculator/PricingCalculatorSample.cls

```diff
@@ -0,0 +1,264 @@
+/**
+* This Pricing calculator extension class makes a call to an external service or static to retrieve pricing
+* information for a cart item and saves it to a cart data transfer object (DTO).
+* For a pricing calculator extension to be processed by the checkout flow, you must implement the
+* CartExtension.PricingCartCalculator class.
+* This is just a sample for Pricing calculator, you are expected to refer this and write your own implementation.
+*/
+public class PricingCalculatorSample extends CartExtension.PricingCartCalculator {
+
+    private static Boolean useExternalService = false;
+
+    public virtual override void calculate(CartExtension.CartCalculateCalculatorRequest request) {
+        try {
+            CartExtension.Cart cart = request.getCart();
+            Iterator<CartExtension.CartItem> cartItemsIterator = clearErrorsAndGetCartItemsIterator(cart, request.getOptionalBuyerActionDetails());
+
+            // Get the SKUs from each cart item that needs price calculations
+            Map<String, CartExtension.CartItem> skuToCartItem = new Map<String, CartExtension.CartItem>();
+            while (cartItemsIterator.hasNext()) {
+                CartExtension.CartItem cartItem = cartItemsIterator.next();
+                skuToCartItem.put(cartItem.getSku(), cartItem);
+            }
+
+            Map<String, PricingDataFromExternalService> pricingDataMap = retrievePriceData(skuToCartItem.keySet());
+
+            if(pricingDataMap == Null) {
+                // No data returned means there is an issue with underlying 3rd party service. Populate generic error message for the Buyer.
+                CartExtension.CartValidationOutput cvo = new CartExtension.CartValidationOutput(
+                CartExtension.CartValidationOutputTypeEnum.PRICING,
+                CartExtension.CartValidationOutputLevelEnum.ERROR);
+                cvo.setMessage('There is no pricing data returned by external service');
+                cart.getCartValidationOutputs().add(cvo);
+                return;
+            }
+            applyPricesToCartItems(cart, skuToCartItem.values().iterator(), pricingDataMap);
+
+        }
+        catch (Exception e) {
+            throw new CalloutException('An integration error occured in COMPUTE_PRICES. Contact your admin.');
```

Comment:
Is the purpose of this exception to halt all CCA execution and not save any in-memory changes to `CartDto` up until this point? If so, you can throw `CartCalculateRuntimeException` here with the same message for the Buyer.

Reply:
done